### ISSUE TITLE
networkd: IPv6AcceptRA default behavior could be documented more clearly

### ISSUE THREAD

**Issue by dseomn** [systemd/systemd#20627]
**systemd version the issue has been seen with**

I noticed it in 241-7~deb10u8, but then I looked at the current source code to see how things work now.

**Used distribution**

Debian GNU/Linux 10 (buster)

**Linux kernel version used** (`uname -a`)

`Linux sakaar 4.19.0-17-amd64 #1 SMP Debian 4.19.194-3 (2021-07-18) x86_64 GNU/Linux`

**CPU architecture issue was seen on**

amd64

**Unexpected behaviour you saw**

I was using sysctl directly to enable forwarding outside of systemd-networkd and I didn't set `IPv6AcceptRA` explicitly. As far as I could tell, systemd-networkd was accepting RAs. From looking at the code, I don't think that's actually a problem, but I think the documentation could be more clear about how `IPv6AcceptRA` works.

The default behavior for `IPv6AcceptRA` is documented to depend on whether "IP forwarding is enabled": https://github.com/systemd/systemd/blob/b2cdc2c05ee77405fa00f3a330bd19cb755a01d0/man/systemd.network.xml#L823-L824

Looking at the code, that seems to mean looking at the `network->ip_forward` value: https://github.com/systemd/systemd/blob/b2cdc2c05ee77405fa00f3a330bd19cb755a01d0/src/network/networkd-ndisc.c#L85-L87

Searching the code, I think the main place that's set is from the configuration item `IPForward`: https://github.com/systemd/systemd/blob/b2cdc2c05ee77405fa00f3a330bd19cb755a01d0/src/network/networkd-network-gperf.gperf#L117

If that's right, would it be possible to clarify the documentation to say that the default behavior of `IPv6AcceptRA` depends on the value of `IPForward` rather than the more general "IP forwarding is enabled"? That would have saved me a lot of effort debugging things if it were clear in the documentation that systemd-networkd isn't checking if forwarding is enabled in the kernel.

**Comment by Ma27** [comment]
I wonder what the motivation behind turning off IPv6AcceptRA with forwarding enabled in the first place?

On NixOS, we have a "fallback" networkd unit that essentially sets `DHCP=yes` for wifi and eth interfaces (with a low "priority" such that anything else would override this). Now, since https://github.com/systemd/systemd/pull/33457 it's advised to turn on IPv6Forward globally rather than per-interface.

This means with globally turning on IPv6Forward, you end up with a machine (given that behavior on NixOS) you end up with a machine that only does DHCPv4 and no SLAAC.

Prompted with this rather surprising behavior, I wonder if that interaction between the two options should be changed entirely?

Related discussion: https://github.com/NixOS/nixpkgs/pull/457644

**Comment by dseomn** [comment]
> I wonder what the motivation behind turning off IPv6AcceptRA with forwarding enabled in the first place?

In general, I'm *guessing* it's just because it's uncommon for routers to have dynamic addresses.

In my case, I looked through my commit history in my personal configs and found https://github.com/dseomn/system-configs/commit/f532dc9ee5f1917ec7c56e46dbb5186009b2fef5. It looks like for the interfaces that I had dnsmasq running on, dnsmasq was sending out RAs and systemd-networkd was reconfiguring the interface based on the RAs that dnsmasq sent over the same interface. And that caused some trouble that I documented in that commit.

**Comment by Ma27** [comment]
> In general, I'm guessing it's just because it's uncommon for routers to have dynamic addresses.

You should still be sufficiently covered by IPv6AcceptRA still being turned off when an interface has IPv6SendRA, no?

**Comment by dseomn** [comment]
That only applies if systemd-networkd is the thing sending RAs, right?

**Comment by yuwata** [comment]
Yeah. I also guess the behavior is for when networkd is running on a router.

BTW, cc7ccd3c6a2be9203057f81be4ef55e94b2ab85c should improve the situation of the original request. Closing.

**Comment by Ma27** [comment]
@yuwata what do you think of my suggestion re. altering the defaults (https://github.com/systemd/systemd/issues/20627#issuecomment-3477892902)?

**Comment by yuwata** [comment]
> [@yuwata](https://github.com/yuwata) what do you think of my suggestion re. altering the defaults ([#20627 (comment)](https://github.com/systemd/systemd/issues/20627#issuecomment-3477892902))?

@Ma27 Please file as a new issue/RFE. Otherwise we easily forget your suggestion. Thanks.
